### DISCUSSION TITLE
Thread safety?

### DISCUSSION THREAD

**Discussion by cocowalla** [devodo/DivertR#62] [Q&A]
Hi, looks interesting! I'm curious about thread safety when used with `WebApplicationFactory`. Take for example one of the code samples from the readme:

```c#
[Fact]
public async Task GivenFooExistsInRepo_WhenGetFoo_ThenReturnsFoo_WithOk200()
{
    // ARRANGE
    var foo = new Foo
    {
        Id = Guid.NewGuid(),
        Name = "Foo123"
    };

    _diverter
        .Redirect<IFooRepository>() // Redirect IFooRepository calls 
        .To(x => x.GetFooAsync(foo.Id)) // matching this method and argument
        .Via(() => Task.FromResult(foo)); // via this delegate

    // ACT
    var response = await _fooClient.GetFooAsync(foo.Id);
    
    // ASSERT
    response.StatusCode.ShouldBe(HttpStatusCode.OK);
    response.Content.Id.ShouldBe(foo.Id);
    response.Content.Name.ShouldBe(foo.Name);
}
```

It replaces uses of `IFooRepository` with something else - but will any other tests running concurrently start using the mocked service too, or does this somehow only affect this one test?

**Comment by devodo** [comment]
Hi and thanks for your question. The services being mocked belong to a WebApplicationFactory instance. So you are correct and you would want to ensure tests run sequentially per instance. The mocked services are reset between each test so each one starts with a clean slate.

Running tests in parallel with xUnit does work well with a separate WebApplicationFactory instance per test collection. From the xUnit docs:

```
By default, each test class is a unique test collection. Tests within the same test class will not run in parallel against each other.
```

The sample test project is using this parallel test collection per class approach:
https://github.com/devodo/DivertR/tree/main/test/DivertR.WebAppTests
